### PR TITLE
feat(tracer): [SVLS-5675] DynamoDB BatchWriteItem pointers

### PR DESCRIPTION
**Pull request by apiarian-datadog**
These are a mix of `PutItem` and `DeleteItem` logic, so they're getting pretty funky.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by apiarian-datadog** [comment]
span links show up as expected in the live test lambda

File: tests/tracer/utils_botocore/test_span_pointers.py [hunk 1]
```diff
@@ -650,3 +755,335 @@
...
+
+            return
+
+        processed_items = _identify_dynamodb_batch_write_item_processed_items(
+            requested_items=test_case.requested_items,
+            unprocessed_items=test_case.unprocessed_items,
+        )
+        assert processed_items == test_case.expected_processed_items
+
+        def collect_all_ids(thing: Any, accumulator: Set[int]) -> None:
```
[the remark this answers is not in the document]
Comment (apiarian-datadog) [reply tests/tracer/utils_botocore/test_span_pointers.py]: fixed

File: tests/tracer/utils_botocore/test_span_pointers.py [hunk 2]
```diff
@@ -650,3 +755,335 @@ def test_pointers(self, pointers_case: PointersCase) -> None:
...
+                for item in thing:
+                    collect_all_ids(item, accumulator)
+
+            elif isinstance(thing, str):
+                # These can be reused internally, but that's fine since they
+                # are immutable.
+                pass
+
+            else:
+                raise Exception(f"unknown type of thing: {type(thing)}")
```
[the remark this answers is not in the document]
Comment (apiarian-datadog) [reply tests/tracer/utils_botocore/test_span_pointers.py]: fixed.

File: ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py [hunk 3]
```diff
@@ -93,6 +125,73 @@
...
+    for table_name, requested_write_requests in requested_items.items():
+        if table_name not in unprocessed_items:
+            processed_items[table_name] = deepcopy(requested_write_requests)
+
+        else:
+            if not all(
+                unprocessed_write_request in requested_write_requests
+                for unprocessed_write_request in unprocessed_items[table_name]
+            ):
+                raise Exception("unprocessed write requests include items not in the requested write requests")
```
[the remark this answers is not in the document]
Comment (apiarian-datadog) [reply ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py]: fixed.

File: ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py [hunk 4]
```diff
@@ -93,6 +125,73 @@ def _extract_span_pointers_for_dynamodb_keyed_operation_response(
...
+
+
+def _identify_dynamodb_batch_write_item_processed_items(
+    requested_items: Dict[_DynamoDBTableName, List[_DynamoDBWriteRequest]],
+    unprocessed_items: Dict[_DynamoDBTableName, List[_DynamoDBWriteRequest]],
+) -> Dict[_DynamoDBTableName, List[_DynamoDBWriteRequest]]:
+    processed_items = {}
+
+    if not all(table_name in requested_items for table_name in unprocessed_items):
+        raise Exception("unprocessed items include tables not in the requested items")
```
[the remark this answers is not in the document]
Comment (apiarian-datadog) [reply ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py]: fixed.

File: releasenotes/notes/span-pointers-aws-dynamodb-batchwriteitem-7c2321a26f212520.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    botocore: Adds span pointers for successful DynamoDB ``BatchWriteItem`` spans. Table Primary Keys will  need to be provided with the ``ddtrace.config.botocore.dynamodb_primary_key_names_for_tables`` option or the ``DD_BOTOCORE_DYNAMODB_TABLE_PRIMARY_KEYS`` environment variable to correctly handdle the ``PutRequest`` items.
```
Comment (ZStriker19) [inline releasenotes/notes/span-pointers-aws-dynamodb-batchwriteitem-7c2321a26f212520.yaml:4]:
```suggestion
    botocore: Adds span pointers for successful DynamoDB ``BatchWriteItem`` spans. Table Primary Keys will need to be provided with the ``ddtrace.config.botocore.dynamodb_primary_key_names_for_tables`` option or the ``DD_BOTOCORE_DYNAMODB_TABLE_PRIMARY_KEYS`` environment variable to correctly handdle the ``PutRequest`` items.
```

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**